File: boards/rectangular.ts
import { BaseBoard, hasInterWall, isEnabled, setInterWallValue } from "../base.ts";
import { PartialExcept } from "../types.ts";
export enum Direction {
    TOP = 0b0001,
    RIGHT = 0b0010,
    BOTTOM = 0b0100,
    LEFT = 0b1000
}
export interface Size {
    height: number;
    width: number;
}
export interface Position {
    x: number;
    y: number;
}
export interface RectangularBoard extends BaseBoard {
    size: Size;
}
export const _size_params = ["height", "width"];
export function newBoard(size: Size): RectangularBoard {
    return {
        cells: new Uint8Array(size.width * size.height),
        size
    };
}
export function newFromBaseBoard({ cells }: BaseBoard, size: Size): RectangularBoard {
    return { cells: cells, size };
}
export function toIndex(position: Position, { size }: PartialExcept<RectangularBoard, 'size'>) {
    return position.y * size.width + position.x;
}
export function toPosition(index: number, { size }: PartialExcept<RectangularBoard, 'size'>) {
    return {
        x: index % size.width,
        y: Math.floor(index / size.width)
    };
}
export function getCellByPosition(position: Position, { cells, size }: RectangularBoard): number {
    return cells[toIndex(position, { size })];
}
export function setCellByPosition(position: Position, value: number, { cells, size }: RectangularBoard) {
    return cells[toIndex(position, { size })] = value;
}
export function getOpposingDirection(direction: Direction): Direction {
    return ((direction << 2) | (direction >> 2)) & 0b1111;
}
export function getRelativeDirection(index1: number, index2: number, { size }: PartialExcept<RectangularBoard, 'size'>): Direction {
    const pos1 = toPosition(index1, { size });
    const pos2 = toPosition(index2, { size });
    if (pos1.y > pos2.y)
        return Direction.TOP;
    if (pos1.x < pos2.x)
        return Direction.RIGHT;
    if (pos1.y < pos2.y)
        return Direction.BOTTOM;
    if (pos1.x > pos2.x)
        return Direction.LEFT;
    throw `'${pos1}' and '${pos2}' are not neighbours`;
}
export function getRelativePosition({ x, y }: Position, direction: Direction) {
    let newPosition = { x, y };
    if (direction === Direction.TOP)
        newPosition.y--;
    if (direction === Direction.RIGHT)
        newPosition.x++;
    if (direction === Direction.BOTTOM)
        newPosition.y++;
    if (direction === Direction.LEFT)
        newPosition.x--;
    return newPosition;
}
export function getNeighbours(index: number, { cells, size }: RectangularBoard): number[] {
    let neighboursCells: number[] = [];
    if (index >= size.width) {
        neighboursCells.push(index - size.width);
    }
    if ((index + 1) % size.width != 0) {
        neighboursCells.push(index + 1);
    }
    if (index < cells.length - size.width) {
        neighboursCells.push(index + size.width);
    }
    if (index % size.width != 0) {
        neighboursCells.push(index - 1);
    }
    neighboursCells = neighboursCells.filter((i) => isEnabled(cells[i]));
    return neighboursCells;
}
export function getAllowedDirection({ x, y }: Position, { cells, size }: RectangularBoard, visitableOnly = true) {
    let directions: Direction[] = [];
    if (y > 0)
        directions.push(Direction.TOP);
    if (x < size.width - 1)
        directions.push(Direction.RIGHT);
    if (y < size.height - 1)
        directions.push(Direction.BOTTOM);
    if (x > 0)
        directions.push(Direction.LEFT);
    directions.filter((dir) => {
        const newPos = getRelativePosition({ x, y }, dir);
        const index1 = toIndex(newPos, { size });
        const index2 = toIndex({ x, y }, { size });
        const cell = cells[index1];
        if (visitableOnly && hasInterWall<RectangularBoard, Direction>(index1, index2, { cells: cells, size }, getRelativeDirection, getOpposingDirection)) {
            return false;
        }
        return isEnabled(cell);
    });
    return directions;
}
export function getRows({ cells, size }: RectangularBoard): number[][] {
    return Array.from(cells)
        .map((_, i) => i)
        .reduce((acc, item, index) => {
        if (index % size.width === 0) {
            acc.push([]);
        }
        acc[acc.length - 1].push(item);
        return acc;
    }, [] as number[][])
        .map((row) => row.filter((c) => isEnabled(cells[c])))
        .filter((row) => row.length);
}
export function removeInterWall(index1: number, index2: number, board: RectangularBoard): RectangularBoard {
    return setInterWallValue<RectangularBoard, Direction>(index1, index2, board, getOpposingDirection, getRelativeDirection, (cell, dir) => cell | dir);
}
export function setInterWall(index1: number, index2: number, board: RectangularBoard): RectangularBoard {
    return setInterWallValue<RectangularBoard, Direction>(index1, index2, board, getOpposingDirection, getRelativeDirection, (cell, dir) => cell & ~dir);
}

File: rust/src/boards/rectangular.rs
use crate::boards::base::BaseBoard;
use crate::{boards::base::BaseDirection, cell::Cell};

#[derive(Copy, Clone)]
pub enum Direction {
    TOP    = 0b0001,
    RIGHT  = 0b0010,
    BOTTOM = 0b0100,
    LEFT   = 0b1000,
}

impl BaseDirection for Direction {
    /// get opposing direction of given wall direction
    /// get_oppositing_wall(Direction.LEFT) -> Direction.RIGHT
    fn get_opposing_dir(&self) -> Self {
        match *self {
            Direction::TOP => Direction::BOTTOM,
            Direction::BOTTOM => Direction::TOP,
            Direction::LEFT => Direction::RIGHT,
            Direction::RIGHT => Direction::LEFT,
        }
    }

    fn get_val(&self) -> u8 {
        return *self as u8;
    }
}

#[derive(Debug)]
pub struct Position {
    x: usize,
    y: usize,
}

pub struct Size {
    width: usize,
    height: usize,
}

pub struct RectangularBoard {
    size: Size,
    cells: Vec<Cell>,
}

impl RectangularBoard {
    // -------------------------
    // Position Functions
    // -------------------------

    /// Linear index from position
    fn to_position(&self, index: usize) -> Position {
        return Position {
            x: index % self.size.width,
            y: (index as f32 / self.size.width as f32).floor() as usize,
        }
    }

    /// Position from linear index
    fn to_index(&self, pos: Position) -> usize {
        return pos.y * self.size.width + pos.x;
    }
}

impl BaseBoard<Direction> for RectangularBoard {
    fn get_nth_cell(&self, index: usize) -> Cell {
        return self.cells[index];
    }

    fn set_nth_cell(&mut self, index: usize, val: u8) {
        self.cells[index].0 = val;
    }

    fn get_relative_direciton(&self, cell_index_1: usize, cell_index_2: usize) -> Direction {
        let pos1 = self.to_position(cell_index_1);
        let pos2 = self.to_position(cell_index_2);

        if pos1.y > pos2.y { return Direction::TOP; }
        if pos1.x < pos2.x { return Direction::RIGHT; }
        if pos1.y < pos2.y { return Direction::BOTTOM; }
        if pos1.x > pos2.x { return Direction::LEFT; }

        panic!("'{:?}' and '{:?}' are not neighbours", pos1, pos2);
    }

    /// return array of rows of cells
    fn get_rows(&self) -> Vec<Vec<Cell>> {
        let mut i= 0;
        let mut rows: Vec<Vec<Cell>> = vec![];

        for _ in 1..self.size.height {
            let mut row: Vec<Cell> = vec![];

            for _1 in 1..self.size.width {
                let cell = self.get_nth_cell(i);
                if cell.is_enabled() { row.extend_from_slice(&[self.get_nth_cell(i)]) }
                i += 1;
            }

            if row.len() > 0 { rows.extend_from_slice(&[row]) }
        }

        return rows;
    }

    /// Get neighbour cells of the given position
    fn get_neighbours(&self, index: usize) -> Vec<Cell> {
        let mut neighbours: Vec<Cell> = vec![];

        // TOP
        if index >= self.size.width {
            let cell = self.get_nth_cell(index - self.size.width);
            if cell.is_enabled() { neighbours.extend_from_slice(&[cell]); }
        }
        // RIGHT
        if (index + 1) % self.size.width != 0 {
            let cell = self.get_nth_cell(index + 1);
            if cell.is_enabled() { neighbours.extend_from_slice(&[cell]); }
        }
        // BOTTOM
        if index < self.cells.len() - self.size.width {
            let cell = self.get_nth_cell(index + self.size.width);
            if cell.is_enabled() { neighbours.extend_from_slice(&[cell]); }
        }
        // LEFT
        if index % self.size.width != 0 {
            let cell = self.get_nth_cell(index - 1);
            if cell.is_enabled() { neighbours.extend_from_slice(&[cell]); }
        }

        return neighbours;
    }
}


#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_direction() {
        assert_eq!(Cell(1).0, 1);
    }
}

